firefox_puppeteer/api/utils.py

```diff
@@ -0,0 +1,76 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from marionette.errors import TimeoutException
+
+from ..base import BaseLib
+
+
+class Utils(BaseLib):
+    """Low-level access to utility actions."""
+
+    def sanitize(self, spec):
+        """Sanitize user data, including cache, cookies, offlineApps, history, formdata,
+        downloads, passwords, sessions, siteSettings.
+
+        Usage:
+        sanitize():  Clears all user data.
+        sanitize({ "sessions": True }): Clears only session user data.
+
+        more: https://dxr.mozilla.org/mozilla-central/source/browser/base/content/sanitize.js
+
+        :param spec: optional, Information specifying data to be sanitized
+        """
+
+        try:
+            self.marionette.execute_async_script("""
+              Cu.import("resource://gre/modules/Services.jsm");
+
+              var aSpec = arguments[0];
+
+              var spec = (typeof aSpec === "undefined") ? {} : {
+                cache: aSpec.cache || false,
+                cookies: aSpec.cookies || false,
+                downloads: aSpec.downloads || false,
+                formdata: aSpec.formdata || false,
+                history: aSpec.history || false,
+                offlineApps: aSpec.offlineApps || false,
+                passwords: aSpec.passwords || false,
+                sessions: aSpec.sessions || false,
+                siteSettings: aSpec.siteSettings || false
+              };
+
+              // Load the sanitize script
+              var tempScope = {};
+              Cc["@mozilla.org/moz/jssubscript-loader;1"]
+              .getService(Ci.mozIJSSubScriptLoader)
+              .loadSubScript("chrome://browser/content/sanitize.js", tempScope);
+
+              // Instantiate the Sanitizer
+              var s = new tempScope.Sanitizer();
+              s.prefDomain = "privacy.cpd.";
+              var itemPrefs = Services.prefs.getBranch(s.prefDomain);
+
+              // Apply options for what to sanitize
+              for (var pref in spec) {
+                itemPrefs.setBoolPref(pref, spec[pref]);
+              };
+
+              // Sanitize and wait for the promise to resolve
+              var finished = false;
+              s.sanitize().then(() => {
+                for (let pref in spec) {
+                  itemPrefs.clearUserPref(pref);
+                };
+                marionetteScriptFinished(true);
+              }, aError => {
+                for (let pref in spec) {
+                  itemPrefs.clearUserPref(pref);
+                };
+                marionetteScriptFinished(false);
+              });
+            """, script_args=[spec], script_timeout=10000)
+        except TimeoutException:
+            # in case of a timeout, continue, since this won't necessarily cause test failure
+            pass
```

Comment:
This would cause a test failure if other tests afterward would rely on a cleaned-up profile. If the max time could even exceed the 5s we might want to set it to 30 but not 10000s, which you did here. I think that Marionette takes it as seconds and not milliseconds.